### PR TITLE
[IR] Add type inference for loop variables

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #3301

### CODE REVIEW

File: tests/python/test_mpm88.py [hunk 1]
```diff
@@ -150,8 +150,7 @@ def substep(x: ti.any_arr(element_dim=1), v: ti.any_arr(element_dim=1),
             fx = x[p] * inv_dx - base.cast(float)
             w = [0.5 * (1.5 - fx)**2, 0.75 - (fx - 1)**2, 0.5 * (fx - 0.5)**2]
             stress = -dt * p_vol * (J[p] - 1) * 4 * inv_dx * inv_dx * E
-            affine = ti.Matrix([[stress, 0], [0, stress]],
-                               dt=ti.f32) + p_mass * C[p]
+            affine = ti.Matrix([[stress, 0], [0, stress]]) + p_mass * C[p]
```
Comment (ailzhang) [inline tests/python/test_mpm88.py:153]: just for my own understanding: is this line related to this PR ?  :P

**Review by ailzhang — APPROVED** [review]
LGTM!

File: tests/python/test_mpm88.py (hunk 1 above)
Comment (strongoier) [reply tests/python/test_mpm88.py:153]: Yep. Removing `dt=ti.f32` means type of `stress` can be inferred.  `stress` relies on `J[p]`, and `p` is a loop variable.

Comment (ailzhang) [reply tests/python/test_mpm88.py:153]: makes sense! thanks!
